In [3]:
import pandas as pd
import json
import configparser
import numpy as np
import os

In [4]:
config_path = '/Users/Jinyue/Documents/go-ethereum/analysis_tool_python/env.conf'
# config_path = '/Users/charlesge/go/src/github.com/ethereum/go-ethereum/analysis_tool_python/env.conf'
config = configparser.ConfigParser()
config.read(config_path)

# we use json paths in this project
UNCLE_PATH = config.get('uncle', 'json_path')
CANONICAL_PATH = config.get('canonical', 'json_path')
FORKED_PATH = config.get('forked', 'json_path')
BROADCAST_ALI_PATH = config.get('broadcast_ali', 'json_path')
BROADCAST_AWS_PATH = config.get('broadcast_aws', 'json_path')

## Load Broadcast 

In [5]:
broadcast_df = None

for filename in sorted(os.listdir(BROADCAST_ALI_PATH)):
    if filename.endswith(".json"):
        print('loading ', filename)
        new_broadcast = pd.read_json(BROADCAST_ALI_PATH + filename, lines=True)
        broadcast_df = pd.concat([broadcast_df, new_broadcast], ignore_index=True)

print(broadcast_df.shape)
broadcast_df.head()


loading  2019-07-09.json
loading  2019-07-10.json
loading  2019-07-11.json
loading  2019-07-12.json
loading  2019-07-13.json
loading  2019-07-14.json
loading  2019-07-15.json
loading  2019-07-16.json
loading  2019-07-17.json
loading  2019-07-18.json
loading  2019-07-19.json
loading  2019-07-20.json
loading  2019-07-21.json
loading  2019-07-22.json
loading  2019-07-23.json
loading  2019-07-24.json
loading  2019-07-25.json
loading  2019-07-26.json
loading  2019-07-27.json
loading  2019-07-28.json
loading  2019-07-29.json
loading  2019-07-30.json
loading  2019-07-31.json
loading  2019-08-01.json
loading  2019-08-02.json
loading  2019-08-03.json
loading  2019-08-09.json


/Users/Jinyue/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:8: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version
of pandas will change to not sort by default.

To accept the future behavior, pass 'sort=False'.

To retain the current behavior and silence the warning, pass 'sort=True'.

  


loading  2019-08-10.json


/Users/Jinyue/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:8: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version
of pandas will change to not sort by default.

To accept the future behavior, pass 'sort=False'.

To retain the current behavior and silence the warning, pass 'sort=True'.

  


loading  2019-08-11.json


/Users/Jinyue/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:8: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version
of pandas will change to not sort by default.

To accept the future behavior, pass 'sort=False'.

To retain the current behavior and silence the warning, pass 'sort=True'.

  


loading  2019-08-12.json


/Users/Jinyue/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:8: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version
of pandas will change to not sort by default.

To accept the future behavior, pass 'sort=False'.

To retain the current behavior and silence the warning, pass 'sort=True'.

  


loading  2019-08-13.json


/Users/Jinyue/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:8: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version
of pandas will change to not sort by default.

To accept the future behavior, pass 'sort=False'.

To retain the current behavior and silence the warning, pass 'sort=True'.

  


loading  2019-08-14.json


/Users/Jinyue/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:8: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version
of pandas will change to not sort by default.

To accept the future behavior, pass 'sort=False'.

To retain the current behavior and silence the warning, pass 'sort=True'.

  


loading  2019-08-15.json


/Users/Jinyue/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:8: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version
of pandas will change to not sort by default.

To accept the future behavior, pass 'sort=False'.

To retain the current behavior and silence the warning, pass 'sort=True'.

  


loading  2019-08-16.json


/Users/Jinyue/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:8: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version
of pandas will change to not sort by default.

To accept the future behavior, pass 'sort=False'.

To retain the current behavior and silence the warning, pass 'sort=True'.

  


loading  2019-08-17.json


/Users/Jinyue/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:8: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version
of pandas will change to not sort by default.

To accept the future behavior, pass 'sort=False'.

To retain the current behavior and silence the warning, pass 'sort=True'.

  


loading  2019-08-18.json


/Users/Jinyue/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:8: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version
of pandas will change to not sort by default.

To accept the future behavior, pass 'sort=False'.

To retain the current behavior and silence the warning, pass 'sort=True'.

  


loading  2019-08-19.json


/Users/Jinyue/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:8: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version
of pandas will change to not sort by default.

To accept the future behavior, pass 'sort=False'.

To retain the current behavior and silence the warning, pass 'sort=True'.

  


loading  2019-08-20.json


/Users/Jinyue/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:8: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version
of pandas will change to not sort by default.

To accept the future behavior, pass 'sort=False'.

To retain the current behavior and silence the warning, pass 'sort=True'.

  


loading  2019-08-21.json


/Users/Jinyue/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:8: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version
of pandas will change to not sort by default.

To accept the future behavior, pass 'sort=False'.

To retain the current behavior and silence the warning, pass 'sort=True'.

  


loading  2019-08-22.json


/Users/Jinyue/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:8: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version
of pandas will change to not sort by default.

To accept the future behavior, pass 'sort=False'.

To retain the current behavior and silence the warning, pass 'sort=True'.

  


loading  2019-08-23.json


/Users/Jinyue/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:8: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version
of pandas will change to not sort by default.

To accept the future behavior, pass 'sort=False'.

To retain the current behavior and silence the warning, pass 'sort=True'.

  


loading  2019-08-24.json


/Users/Jinyue/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:8: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version
of pandas will change to not sort by default.

To accept the future behavior, pass 'sort=False'.

To retain the current behavior and silence the warning, pass 'sort=True'.

  


loading  2019-08-25.json


/Users/Jinyue/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:8: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version
of pandas will change to not sort by default.

To accept the future behavior, pass 'sort=False'.

To retain the current behavior and silence the warning, pass 'sort=True'.

  


loading  2019-08-26.json


/Users/Jinyue/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:8: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version
of pandas will change to not sort by default.

To accept the future behavior, pass 'sort=False'.

To retain the current behavior and silence the warning, pass 'sort=True'.

  


loading  2019-08-27.json


/Users/Jinyue/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:8: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version
of pandas will change to not sort by default.

To accept the future behavior, pass 'sort=False'.

To retain the current behavior and silence the warning, pass 'sort=True'.

  


loading  2019-08-28.json


/Users/Jinyue/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:8: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version
of pandas will change to not sort by default.

To accept the future behavior, pass 'sort=False'.

To retain the current behavior and silence the warning, pass 'sort=True'.

  


loading  2019-08-29.json


/Users/Jinyue/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:8: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version
of pandas will change to not sort by default.

To accept the future behavior, pass 'sort=False'.

To retain the current behavior and silence the warning, pass 'sort=True'.

  


(814687, 32)


difficulty extra extraData  gasLimit   gasUsed  \
0  2119842492649275   NaN            0x7a20b0  0x484e2d   
1  2119842492649275   NaN            0x7a20b0  0x484e2d   
2  2119843029520187   NaN            0x7a1d46  0x39bfb3   
3  2119843029520187   NaN            0x7a1d46  0x39bfb3   
4                     NaN            0x7a1d46  0x39bfb3   

                                                hash  \
0  0x869d968cbd5404f81d8fa750eb4672f4752db1979474...   
1  0x869d968cbd5404f81d8fa750eb4672f4752db1979474...   
2  0x86c7e5fd62d71430be8292ac909df6a1b3730aae7e07...   
3  0x86c7e5fd62d71430be8292ac909df6a1b3730aae7e07...   
4  0x86c7e5fd62d71430be8292ac909df6a1b3730aae7e07...   

                                           logsBloom mineTime  \
0  0221218a04322008408904084268000981003008400000...            
1  0221218a04322008408904084268000981003008400000...            
2  000050000000002a845c00002408901421c230c2011020...            
3  000050000000002a845c00002408901421c230c2011020...            
4  000050000000002a845c00002408901421c230c2011020...            

                                        miner minerName  ... stateRoot  \
0  0x5A0b54D5dc17e0AadC383d2db43B0a0D3E029c4c            ...             
1  0x5A0b54D5dc17e0AadC383d2db43B0a0D3E029c4c            ...             
2  0x5A0b54D5dc17e0AadC383d2db43B0a0D3E029c4c            ...             
3  0x5A0b54D5dc17e0AadC383d2db43B0a0D3E029c4c            ...             
4  0x5A0b54D5dc17e0AadC383d2db43B0a0D3E029c4c            ...             

            timestamp  totalDifficulty  transactions transactionsRoot txNum  \
0 2019-07-08 22:59:50                                                    65   
1 2019-07-08 22:59:50                                                    65   
2 2019-07-08 22:59:59                                                    58   
3 2019-07-08 22:59:59                                                    58   
4 2019-07-08 22:59:59                                                    58   

  uncleHash uncleNum uncleReward  \
0       NaN      0.0               
1       NaN      0.0               
2       NaN      0.0               
3       NaN      0.0               
4       NaN      0.0               

                                              uncles  
0  [0x1dcc4de8dec75d7aab85b567b6ccd41ad312451b948...  
1  [0x1dcc4de8dec75d7aab85b567b6ccd41ad312451b948...  
2  [0x1dcc4de8dec75d7aab85b567b6ccd41ad312451b948...  
3  [0x1dcc4de8dec75d7aab85b567b6ccd41ad312451b948...  
4  [0x1dcc4de8dec75d7aab85b567b6ccd41ad312451b948...  

[5 rows x 32 columns]

In [8]:
for filename in sorted(os.listdir(BROADCAST_AWS_PATH)):
    if filename.endswith(".json"):
        print('loading ', filename)
        new_broadcast = pd.read_json(BROADCAST_AWS_PATH + filename, lines=True)
        broadcast_df = pd.concat([broadcast_df, new_broadcast], ignore_index=True)

print(broadcast_df.shape)
broadcast_df.head()

loading  2019-07-09.json


/Users/Jinyue/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:5: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version
of pandas will change to not sort by default.

To accept the future behavior, pass 'sort=False'.

To retain the current behavior and silence the warning, pass 'sort=True'.

  """


loading  2019-07-10.json


/Users/Jinyue/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:5: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version
of pandas will change to not sort by default.

To accept the future behavior, pass 'sort=False'.

To retain the current behavior and silence the warning, pass 'sort=True'.

  """


loading  2019-07-13.json


/Users/Jinyue/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:5: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version
of pandas will change to not sort by default.

To accept the future behavior, pass 'sort=False'.

To retain the current behavior and silence the warning, pass 'sort=True'.

  """


loading  2019-07-14.json


/Users/Jinyue/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:5: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version
of pandas will change to not sort by default.

To accept the future behavior, pass 'sort=False'.

To retain the current behavior and silence the warning, pass 'sort=True'.

  """


loading  2019-07-15.json


/Users/Jinyue/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:5: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version
of pandas will change to not sort by default.

To accept the future behavior, pass 'sort=False'.

To retain the current behavior and silence the warning, pass 'sort=True'.

  """


loading  2019-07-16.json


/Users/Jinyue/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:5: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version
of pandas will change to not sort by default.

To accept the future behavior, pass 'sort=False'.

To retain the current behavior and silence the warning, pass 'sort=True'.

  """


loading  2019-07-17.json


/Users/Jinyue/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:5: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version
of pandas will change to not sort by default.

To accept the future behavior, pass 'sort=False'.

To retain the current behavior and silence the warning, pass 'sort=True'.

  """


loading  2019-07-18.json


/Users/Jinyue/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:5: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version
of pandas will change to not sort by default.

To accept the future behavior, pass 'sort=False'.

To retain the current behavior and silence the warning, pass 'sort=True'.

  """


loading  2019-07-19.json


/Users/Jinyue/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:5: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version
of pandas will change to not sort by default.

To accept the future behavior, pass 'sort=False'.

To retain the current behavior and silence the warning, pass 'sort=True'.

  """


loading  2019-07-20.json


/Users/Jinyue/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:5: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version
of pandas will change to not sort by default.

To accept the future behavior, pass 'sort=False'.

To retain the current behavior and silence the warning, pass 'sort=True'.

  """


loading  2019-07-21.json


/Users/Jinyue/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:5: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version
of pandas will change to not sort by default.

To accept the future behavior, pass 'sort=False'.

To retain the current behavior and silence the warning, pass 'sort=True'.

  """


loading  2019-07-22.json


/Users/Jinyue/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:5: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version
of pandas will change to not sort by default.

To accept the future behavior, pass 'sort=False'.

To retain the current behavior and silence the warning, pass 'sort=True'.

  """


loading  2019-07-23.json


/Users/Jinyue/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:5: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version
of pandas will change to not sort by default.

To accept the future behavior, pass 'sort=False'.

To retain the current behavior and silence the warning, pass 'sort=True'.

  """


loading  2019-07-24.json


/Users/Jinyue/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:5: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version
of pandas will change to not sort by default.

To accept the future behavior, pass 'sort=False'.

To retain the current behavior and silence the warning, pass 'sort=True'.

  """


loading  2019-07-25.json


/Users/Jinyue/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:5: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version
of pandas will change to not sort by default.

To accept the future behavior, pass 'sort=False'.

To retain the current behavior and silence the warning, pass 'sort=True'.

  """


loading  2019-07-29.json


/Users/Jinyue/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:5: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version
of pandas will change to not sort by default.

To accept the future behavior, pass 'sort=False'.

To retain the current behavior and silence the warning, pass 'sort=True'.

  """


loading  2019-07-30.json


/Users/Jinyue/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:5: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version
of pandas will change to not sort by default.

To accept the future behavior, pass 'sort=False'.

To retain the current behavior and silence the warning, pass 'sort=True'.

  """


loading  2019-07-31.json


/Users/Jinyue/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:5: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version
of pandas will change to not sort by default.

To accept the future behavior, pass 'sort=False'.

To retain the current behavior and silence the warning, pass 'sort=True'.

  """


loading  2019-08-01.json


/Users/Jinyue/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:5: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version
of pandas will change to not sort by default.

To accept the future behavior, pass 'sort=False'.

To retain the current behavior and silence the warning, pass 'sort=True'.

  """


loading  2019-08-02.json


/Users/Jinyue/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:5: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version
of pandas will change to not sort by default.

To accept the future behavior, pass 'sort=False'.

To retain the current behavior and silence the warning, pass 'sort=True'.

  """


loading  2019-08-03.json


/Users/Jinyue/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:5: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version
of pandas will change to not sort by default.

To accept the future behavior, pass 'sort=False'.

To retain the current behavior and silence the warning, pass 'sort=True'.

  """


loading  2019-08-04.json


/Users/Jinyue/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:5: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version
of pandas will change to not sort by default.

To accept the future behavior, pass 'sort=False'.

To retain the current behavior and silence the warning, pass 'sort=True'.

  """


loading  2019-08-05.json


/Users/Jinyue/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:5: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version
of pandas will change to not sort by default.

To accept the future behavior, pass 'sort=False'.

To retain the current behavior and silence the warning, pass 'sort=True'.

  """


loading  2019-08-06.json


/Users/Jinyue/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:5: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version
of pandas will change to not sort by default.

To accept the future behavior, pass 'sort=False'.

To retain the current behavior and silence the warning, pass 'sort=True'.

  """


loading  2019-08-09.json


/Users/Jinyue/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:5: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version
of pandas will change to not sort by default.

To accept the future behavior, pass 'sort=False'.

To retain the current behavior and silence the warning, pass 'sort=True'.

  """


loading  2019-08-10.json


/Users/Jinyue/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:5: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version
of pandas will change to not sort by default.

To accept the future behavior, pass 'sort=False'.

To retain the current behavior and silence the warning, pass 'sort=True'.

  """


loading  2019-08-11.json


/Users/Jinyue/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:5: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version
of pandas will change to not sort by default.

To accept the future behavior, pass 'sort=False'.

To retain the current behavior and silence the warning, pass 'sort=True'.

  """


loading  2019-08-12.json


/Users/Jinyue/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:5: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version
of pandas will change to not sort by default.

To accept the future behavior, pass 'sort=False'.

To retain the current behavior and silence the warning, pass 'sort=True'.

  """


loading  2019-08-13.json


/Users/Jinyue/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:5: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version
of pandas will change to not sort by default.

To accept the future behavior, pass 'sort=False'.

To retain the current behavior and silence the warning, pass 'sort=True'.

  """


loading  2019-08-14.json


/Users/Jinyue/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:5: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version
of pandas will change to not sort by default.

To accept the future behavior, pass 'sort=False'.

To retain the current behavior and silence the warning, pass 'sort=True'.

  """


loading  2019-08-15.json


/Users/Jinyue/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:5: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version
of pandas will change to not sort by default.

To accept the future behavior, pass 'sort=False'.

To retain the current behavior and silence the warning, pass 'sort=True'.

  """


loading  2019-08-16.json


/Users/Jinyue/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:5: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version
of pandas will change to not sort by default.

To accept the future behavior, pass 'sort=False'.

To retain the current behavior and silence the warning, pass 'sort=True'.

  """


loading  2019-08-17.json


/Users/Jinyue/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:5: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version
of pandas will change to not sort by default.

To accept the future behavior, pass 'sort=False'.

To retain the current behavior and silence the warning, pass 'sort=True'.

  """


loading  2019-08-18.json


/Users/Jinyue/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:5: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version
of pandas will change to not sort by default.

To accept the future behavior, pass 'sort=False'.

To retain the current behavior and silence the warning, pass 'sort=True'.

  """


loading  2019-08-19.json


/Users/Jinyue/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:5: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version
of pandas will change to not sort by default.

To accept the future behavior, pass 'sort=False'.

To retain the current behavior and silence the warning, pass 'sort=True'.

  """


loading  2019-08-20.json


/Users/Jinyue/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:5: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version
of pandas will change to not sort by default.

To accept the future behavior, pass 'sort=False'.

To retain the current behavior and silence the warning, pass 'sort=True'.

  """


loading  2019-08-21.json


/Users/Jinyue/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:5: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version
of pandas will change to not sort by default.

To accept the future behavior, pass 'sort=False'.

To retain the current behavior and silence the warning, pass 'sort=True'.

  """


loading  2019-08-22.json


/Users/Jinyue/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:5: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version
of pandas will change to not sort by default.

To accept the future behavior, pass 'sort=False'.

To retain the current behavior and silence the warning, pass 'sort=True'.

  """


loading  2019-08-23.json


/Users/Jinyue/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:5: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version
of pandas will change to not sort by default.

To accept the future behavior, pass 'sort=False'.

To retain the current behavior and silence the warning, pass 'sort=True'.

  """


loading  2019-08-24.json


/Users/Jinyue/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:5: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version
of pandas will change to not sort by default.

To accept the future behavior, pass 'sort=False'.

To retain the current behavior and silence the warning, pass 'sort=True'.

  """


loading  2019-08-25.json


/Users/Jinyue/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:5: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version
of pandas will change to not sort by default.

To accept the future behavior, pass 'sort=False'.

To retain the current behavior and silence the warning, pass 'sort=True'.

  """


loading  2019-08-26.json


/Users/Jinyue/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:5: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version
of pandas will change to not sort by default.

To accept the future behavior, pass 'sort=False'.

To retain the current behavior and silence the warning, pass 'sort=True'.

  """


loading  2019-08-27.json


/Users/Jinyue/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:5: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version
of pandas will change to not sort by default.

To accept the future behavior, pass 'sort=False'.

To retain the current behavior and silence the warning, pass 'sort=True'.

  """


loading  2019-08-28.json


/Users/Jinyue/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:5: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version
of pandas will change to not sort by default.

To accept the future behavior, pass 'sort=False'.

To retain the current behavior and silence the warning, pass 'sort=True'.

  """


(2752419, 32)


difficulty extra extraData  gasLimit   gasUsed  \
0  2119842492649275   NaN            0x7a20b0  0x484e2d   
1  2119842492649275   NaN            0x7a20b0  0x484e2d   
2  2119843029520187   NaN            0x7a1d46  0x39bfb3   
3  2119843029520187   NaN            0x7a1d46  0x39bfb3   
4                     NaN            0x7a1d46  0x39bfb3   

                                                hash  \
0  0x869d968cbd5404f81d8fa750eb4672f4752db1979474...   
1  0x869d968cbd5404f81d8fa750eb4672f4752db1979474...   
2  0x86c7e5fd62d71430be8292ac909df6a1b3730aae7e07...   
3  0x86c7e5fd62d71430be8292ac909df6a1b3730aae7e07...   
4  0x86c7e5fd62d71430be8292ac909df6a1b3730aae7e07...   

                                           logsBloom mineTime  \
0  0221218a04322008408904084268000981003008400000...            
1  0221218a04322008408904084268000981003008400000...            
2  000050000000002a845c00002408901421c230c2011020...            
3  000050000000002a845c00002408901421c230c2011020...            
4  000050000000002a845c00002408901421c230c2011020...            

                                        miner minerName  ... stateRoot  \
0  0x5A0b54D5dc17e0AadC383d2db43B0a0D3E029c4c            ...             
1  0x5A0b54D5dc17e0AadC383d2db43B0a0D3E029c4c            ...             
2  0x5A0b54D5dc17e0AadC383d2db43B0a0D3E029c4c            ...             
3  0x5A0b54D5dc17e0AadC383d2db43B0a0D3E029c4c            ...             
4  0x5A0b54D5dc17e0AadC383d2db43B0a0D3E029c4c            ...             

            timestamp  totalDifficulty  transactions transactionsRoot txNum  \
0 2019-07-08 22:59:50                                                    65   
1 2019-07-08 22:59:50                                                    65   
2 2019-07-08 22:59:59                                                    58   
3 2019-07-08 22:59:59                                                    58   
4 2019-07-08 22:59:59                                                    58   

  uncleHash uncleNum uncleReward  \
0       NaN      0.0               
1       NaN      0.0               
2       NaN      0.0               
3       NaN      0.0               
4       NaN      0.0               

                                              uncles  
0  [0x1dcc4de8dec75d7aab85b567b6ccd41ad312451b948...  
1  [0x1dcc4de8dec75d7aab85b567b6ccd41ad312451b948...  
2  [0x1dcc4de8dec75d7aab85b567b6ccd41ad312451b948...  
3  [0x1dcc4de8dec75d7aab85b567b6ccd41ad312451b948...  
4  [0x1dcc4de8dec75d7aab85b567b6ccd41ad312451b948...  

[5 rows x 32 columns]

## Load Uncles

In [9]:
uncle_df = None

for filename in sorted(os.listdir(UNCLE_PATH)):
    if filename.endswith(".json"):
        print('loading ', filename)
        new_uncles = pd.read_json(UNCLE_PATH + filename, lines=True)
        uncle_df = pd.concat([uncle_df, new_uncles], ignore_index=True)

print(uncle_df.shape)
uncle_df.head()

loading  7200000.json
loading  7210000.json
loading  7220000.json
loading  7230000.json
loading  7240000.json
loading  7250000.json
loading  7260000.json
loading  7270000.json
loading  7280000.json
loading  7290000.json
loading  7300000.json
loading  7310000.json
loading  7320000.json
loading  7330000.json
loading  7340000.json
loading  7350000.json
loading  7360000.json
loading  7370000.json
loading  7380000.json
loading  7390000.json
loading  7400000.json
loading  7410000.json
loading  7420000.json
loading  7430000.json
loading  7440000.json
loading  7450000.json
loading  7460000.json
loading  7470000.json
loading  7480000.json
loading  7490000.json
loading  7500000.json
loading  7510000.json
loading  7520000.json
loading  7530000.json
loading  7540000.json
loading  7550000.json
loading  7560000.json
loading  7570000.json
loading  7580000.json
loading  7590000.json
loading  7600000.json
loading  7610000.json
loading  7620000.json
loading  7630000.json
loading  7640000.json
loading  7

difficulty                                   extraData  gasLimit  \
0  0x8886feaa77f19                                          0x   7992222   
1  0x88cd0ecd9f73e  0x737061726b706f6f6c2d6574682d636e2d687a32   8000029   
2  0x895a97bdabd2c        0x65746865726d696e652d61736961312d31   8000029   
3  0x8a7418c515067    0xe4b883e5bda9e7a59ee4bb99e9b1bcbf84ab69   8000025   
4  0x8a7418c515067    0xe4b883e5bda9e7a59ee4bb99e9b1bcbf84ab69   8000025   

   gasUsed                                               hash  \
0  5902995  0x26e21747dfd05acbb0db763447dc730dd1d20ddb70a6...   
1  7994199  0x4e440909390986585b35a53b31d285c95899710b9ba6...   
2  5743895  0x1e48f03f15d604bcecade135e218821d8d3c243f7a6a...   
3  7980364  0x8d9a618adde95515b851c248d8a94b73e766af5ea184...   
4  7980364  0x8d9a618adde95515b851c248d8a94b73e766af5ea184...   

                                           logsBloom mineTime  \
0  0x00801010000000010040000188500000002038004108...            
1  0x0c0b14934a0221a50232762d70820048360180288130...            
2  0x080000800143450122230420e00400086110a0280030...            
3  0x0402000000020004000a042000000200200000040000...            
4  0x0402000000020004000a042000000200200000040000...            

                                        miner minerName  \
0  0xcc16e3c00dbbe76603fa833ec20a48f786dfe610             
1  0x5a0b54d5dc17e0aadc383d2db43b0a0d3e029c4c             
2  0xea674fdde714fd979de3edf0f56aa9716b898ec8             
3  0x829bd824b016326a401d083b33d092293333a830             
4  0x829bd824b016326a401d083b33d092293333a830             

                                             mixHash  ... size  \
0  0x45214d0508d89bfedbce676ed92f5df8e6cba231cd38...  ...  521   
1  0xe3f1fb985d20028b78c1f800a71116eea0d6f58ae7b6...  ...  541   
2  0xf4e4a6fae64b84f40485b04107c4ed67a5af27e0d7c9...  ...  538   
3  0x5879fc4afb2e84085bffdb0ae2f6b4de07576bcc8017...  ...  540   
4  0x5879fc4afb2e84085bffdb0ae2f6b4de07576bcc8017...  ...  540   

                                           stateRoot           timestamp  \
0  0x72bb0a8ae404dbee3b76970f275990101ff3f588b7d5... 2019-02-09 22:46:31   
1  0xb38534dd8cb9c19747c8a921bf1ccb504dc0095446b8... 2019-02-09 22:51:03   
2  0xd230d4adeb83a207724cb4b9572d3c70b4ff0c9df6fd... 2019-02-09 22:53:46   
3  0x11ff3fece7d33709ef7aedbfe9e22d79018983a79bfb... 2019-02-09 23:05:16   
4  0x11ff3fece7d33709ef7aedbfe9e22d79018983a79bfb... 2019-02-09 23:05:16   

  totalDifficulty transactions  \
0             NaN           []   
1             NaN           []   
2             NaN           []   
3             NaN           []   
4             NaN           []   

                                    transactionsRoot txNum  uncleNum  \
0  0x6ad23338c6165431390ff61714733f5a7b5054779db6...     0         0   
1  0xbd327fac1bad5a8169651f491cf1b5b6518829091384...     0         0   
2  0xdfa6a0c596d983b4f8caaa22819e9514a5dd6ea00a1e...     0         0   
3  0x27c63a78c0e7cd14c7a1d7d887a2ea72f050aa68bdc6...     0         0   
4  0x27c63a78c0e7cd14c7a1d7d887a2ea72f050aa68bdc6...     0         0   

  uncleReward uncles  
0                 []  
1                 []  
2                 []  
3                 []  
4                 []  

[5 rows x 27 columns]

## Load Forked

In [10]:
forked_df = None

for filename in sorted(os.listdir(FORKED_PATH)):
    if filename.endswith(".txt"):
        print('loading ', FORKED_PATH + filename)
        new_forked = pd.read_json(FORKED_PATH + filename, lines=True)
        forked_df = pd.concat([forked_df, new_forked], ignore_index=True)

#print(forked_df.shape)
forked_df.head()

loading  /Users/Jinyue/Documents/go-ethereum/analysis_tool_python/crawler/forked_json/2016-08-21.txt
loading  /Users/Jinyue/Documents/go-ethereum/analysis_tool_python/crawler/forked_json/2016-08-22.txt
loading  /Users/Jinyue/Documents/go-ethereum/analysis_tool_python/crawler/forked_json/2016-08-23.txt
loading  /Users/Jinyue/Documents/go-ethereum/analysis_tool_python/crawler/forked_json/2016-08-24.txt
loading  /Users/Jinyue/Documents/go-ethereum/analysis_tool_python/crawler/forked_json/2016-08-25.txt
loading  /Users/Jinyue/Documents/go-ethereum/analysis_tool_python/crawler/forked_json/2016-08-26.txt
loading  /Users/Jinyue/Documents/go-ethereum/analysis_tool_python/crawler/forked_json/2016-08-27.txt
loading  /Users/Jinyue/Documents/go-ethereum/analysis_tool_python/crawler/forked_json/2016-08-28.txt
loading  /Users/Jinyue/Documents/go-ethereum/analysis_tool_python/crawler/forked_json/2016-08-29.txt
loading  /Users/Jinyue/Documents/go-ethereum/analysis_tool_python/crawler/forked_json/2016-

loading  /Users/Jinyue/Documents/go-ethereum/analysis_tool_python/crawler/forked_json/2016-11-16.txt
loading  /Users/Jinyue/Documents/go-ethereum/analysis_tool_python/crawler/forked_json/2016-11-17.txt
loading  /Users/Jinyue/Documents/go-ethereum/analysis_tool_python/crawler/forked_json/2016-11-18.txt
loading  /Users/Jinyue/Documents/go-ethereum/analysis_tool_python/crawler/forked_json/2016-11-19.txt
loading  /Users/Jinyue/Documents/go-ethereum/analysis_tool_python/crawler/forked_json/2016-11-20.txt
loading  /Users/Jinyue/Documents/go-ethereum/analysis_tool_python/crawler/forked_json/2016-11-21.txt
loading  /Users/Jinyue/Documents/go-ethereum/analysis_tool_python/crawler/forked_json/2016-11-22.txt
loading  /Users/Jinyue/Documents/go-ethereum/analysis_tool_python/crawler/forked_json/2016-11-23.txt
loading  /Users/Jinyue/Documents/go-ethereum/analysis_tool_python/crawler/forked_json/2016-11-24.txt
loading  /Users/Jinyue/Documents/go-ethereum/analysis_tool_python/crawler/forked_json/2016-

loading  /Users/Jinyue/Documents/go-ethereum/analysis_tool_python/crawler/forked_json/2017-02-07.txt
loading  /Users/Jinyue/Documents/go-ethereum/analysis_tool_python/crawler/forked_json/2017-02-08.txt
loading  /Users/Jinyue/Documents/go-ethereum/analysis_tool_python/crawler/forked_json/2017-02-09.txt
loading  /Users/Jinyue/Documents/go-ethereum/analysis_tool_python/crawler/forked_json/2017-02-10.txt
loading  /Users/Jinyue/Documents/go-ethereum/analysis_tool_python/crawler/forked_json/2017-02-11.txt
loading  /Users/Jinyue/Documents/go-ethereum/analysis_tool_python/crawler/forked_json/2017-02-12.txt
loading  /Users/Jinyue/Documents/go-ethereum/analysis_tool_python/crawler/forked_json/2017-02-13.txt
loading  /Users/Jinyue/Documents/go-ethereum/analysis_tool_python/crawler/forked_json/2017-02-14.txt
loading  /Users/Jinyue/Documents/go-ethereum/analysis_tool_python/crawler/forked_json/2017-02-15.txt
loading  /Users/Jinyue/Documents/go-ethereum/analysis_tool_python/crawler/forked_json/2017-

loading  /Users/Jinyue/Documents/go-ethereum/analysis_tool_python/crawler/forked_json/2017-05-02.txt
loading  /Users/Jinyue/Documents/go-ethereum/analysis_tool_python/crawler/forked_json/2017-05-03.txt
loading  /Users/Jinyue/Documents/go-ethereum/analysis_tool_python/crawler/forked_json/2017-05-04.txt
loading  /Users/Jinyue/Documents/go-ethereum/analysis_tool_python/crawler/forked_json/2017-05-05.txt
loading  /Users/Jinyue/Documents/go-ethereum/analysis_tool_python/crawler/forked_json/2017-05-06.txt
loading  /Users/Jinyue/Documents/go-ethereum/analysis_tool_python/crawler/forked_json/2017-05-07.txt
loading  /Users/Jinyue/Documents/go-ethereum/analysis_tool_python/crawler/forked_json/2017-05-08.txt
loading  /Users/Jinyue/Documents/go-ethereum/analysis_tool_python/crawler/forked_json/2017-05-09.txt
loading  /Users/Jinyue/Documents/go-ethereum/analysis_tool_python/crawler/forked_json/2017-05-10.txt
loading  /Users/Jinyue/Documents/go-ethereum/analysis_tool_python/crawler/forked_json/2017-

loading  /Users/Jinyue/Documents/go-ethereum/analysis_tool_python/crawler/forked_json/2017-07-23.txt
loading  /Users/Jinyue/Documents/go-ethereum/analysis_tool_python/crawler/forked_json/2017-07-24.txt
loading  /Users/Jinyue/Documents/go-ethereum/analysis_tool_python/crawler/forked_json/2017-07-25.txt
loading  /Users/Jinyue/Documents/go-ethereum/analysis_tool_python/crawler/forked_json/2017-07-26.txt
loading  /Users/Jinyue/Documents/go-ethereum/analysis_tool_python/crawler/forked_json/2017-07-27.txt
loading  /Users/Jinyue/Documents/go-ethereum/analysis_tool_python/crawler/forked_json/2017-07-28.txt
loading  /Users/Jinyue/Documents/go-ethereum/analysis_tool_python/crawler/forked_json/2017-07-29.txt
loading  /Users/Jinyue/Documents/go-ethereum/analysis_tool_python/crawler/forked_json/2017-07-30.txt
loading  /Users/Jinyue/Documents/go-ethereum/analysis_tool_python/crawler/forked_json/2017-07-31.txt
loading  /Users/Jinyue/Documents/go-ethereum/analysis_tool_python/crawler/forked_json/2017-

loading  /Users/Jinyue/Documents/go-ethereum/analysis_tool_python/crawler/forked_json/2017-10-15.txt
loading  /Users/Jinyue/Documents/go-ethereum/analysis_tool_python/crawler/forked_json/2017-10-16.txt
loading  /Users/Jinyue/Documents/go-ethereum/analysis_tool_python/crawler/forked_json/2017-10-17.txt
loading  /Users/Jinyue/Documents/go-ethereum/analysis_tool_python/crawler/forked_json/2017-10-18.txt
loading  /Users/Jinyue/Documents/go-ethereum/analysis_tool_python/crawler/forked_json/2017-10-19.txt
loading  /Users/Jinyue/Documents/go-ethereum/analysis_tool_python/crawler/forked_json/2017-10-20.txt
loading  /Users/Jinyue/Documents/go-ethereum/analysis_tool_python/crawler/forked_json/2017-10-21.txt
loading  /Users/Jinyue/Documents/go-ethereum/analysis_tool_python/crawler/forked_json/2017-10-22.txt
loading  /Users/Jinyue/Documents/go-ethereum/analysis_tool_python/crawler/forked_json/2017-10-23.txt
loading  /Users/Jinyue/Documents/go-ethereum/analysis_tool_python/crawler/forked_json/2017-

loading  /Users/Jinyue/Documents/go-ethereum/analysis_tool_python/crawler/forked_json/2018-01-07.txt
loading  /Users/Jinyue/Documents/go-ethereum/analysis_tool_python/crawler/forked_json/2018-01-08.txt
loading  /Users/Jinyue/Documents/go-ethereum/analysis_tool_python/crawler/forked_json/2018-01-09.txt
loading  /Users/Jinyue/Documents/go-ethereum/analysis_tool_python/crawler/forked_json/2018-01-10.txt
loading  /Users/Jinyue/Documents/go-ethereum/analysis_tool_python/crawler/forked_json/2018-01-11.txt
loading  /Users/Jinyue/Documents/go-ethereum/analysis_tool_python/crawler/forked_json/2018-01-12.txt
loading  /Users/Jinyue/Documents/go-ethereum/analysis_tool_python/crawler/forked_json/2018-01-13.txt
loading  /Users/Jinyue/Documents/go-ethereum/analysis_tool_python/crawler/forked_json/2018-01-14.txt
loading  /Users/Jinyue/Documents/go-ethereum/analysis_tool_python/crawler/forked_json/2018-01-15.txt
loading  /Users/Jinyue/Documents/go-ethereum/analysis_tool_python/crawler/forked_json/2018-

loading  /Users/Jinyue/Documents/go-ethereum/analysis_tool_python/crawler/forked_json/2018-03-30.txt
loading  /Users/Jinyue/Documents/go-ethereum/analysis_tool_python/crawler/forked_json/2018-03-31.txt
loading  /Users/Jinyue/Documents/go-ethereum/analysis_tool_python/crawler/forked_json/2018-04-01.txt
loading  /Users/Jinyue/Documents/go-ethereum/analysis_tool_python/crawler/forked_json/2018-04-02.txt
loading  /Users/Jinyue/Documents/go-ethereum/analysis_tool_python/crawler/forked_json/2018-04-03.txt
loading  /Users/Jinyue/Documents/go-ethereum/analysis_tool_python/crawler/forked_json/2018-04-04.txt
loading  /Users/Jinyue/Documents/go-ethereum/analysis_tool_python/crawler/forked_json/2018-04-05.txt
loading  /Users/Jinyue/Documents/go-ethereum/analysis_tool_python/crawler/forked_json/2018-04-06.txt
loading  /Users/Jinyue/Documents/go-ethereum/analysis_tool_python/crawler/forked_json/2018-04-07.txt
loading  /Users/Jinyue/Documents/go-ethereum/analysis_tool_python/crawler/forked_json/2018-

loading  /Users/Jinyue/Documents/go-ethereum/analysis_tool_python/crawler/forked_json/2018-06-20.txt
loading  /Users/Jinyue/Documents/go-ethereum/analysis_tool_python/crawler/forked_json/2018-06-21.txt
loading  /Users/Jinyue/Documents/go-ethereum/analysis_tool_python/crawler/forked_json/2018-06-22.txt
loading  /Users/Jinyue/Documents/go-ethereum/analysis_tool_python/crawler/forked_json/2018-06-23.txt
loading  /Users/Jinyue/Documents/go-ethereum/analysis_tool_python/crawler/forked_json/2018-06-24.txt
loading  /Users/Jinyue/Documents/go-ethereum/analysis_tool_python/crawler/forked_json/2018-06-25.txt
loading  /Users/Jinyue/Documents/go-ethereum/analysis_tool_python/crawler/forked_json/2018-06-26.txt
loading  /Users/Jinyue/Documents/go-ethereum/analysis_tool_python/crawler/forked_json/2018-06-27.txt
loading  /Users/Jinyue/Documents/go-ethereum/analysis_tool_python/crawler/forked_json/2018-06-28.txt
loading  /Users/Jinyue/Documents/go-ethereum/analysis_tool_python/crawler/forked_json/2018-

loading  /Users/Jinyue/Documents/go-ethereum/analysis_tool_python/crawler/forked_json/2018-09-10.txt
loading  /Users/Jinyue/Documents/go-ethereum/analysis_tool_python/crawler/forked_json/2018-09-11.txt
loading  /Users/Jinyue/Documents/go-ethereum/analysis_tool_python/crawler/forked_json/2018-09-12.txt
loading  /Users/Jinyue/Documents/go-ethereum/analysis_tool_python/crawler/forked_json/2018-09-13.txt
loading  /Users/Jinyue/Documents/go-ethereum/analysis_tool_python/crawler/forked_json/2018-09-14.txt
loading  /Users/Jinyue/Documents/go-ethereum/analysis_tool_python/crawler/forked_json/2018-09-15.txt
loading  /Users/Jinyue/Documents/go-ethereum/analysis_tool_python/crawler/forked_json/2018-09-16.txt
loading  /Users/Jinyue/Documents/go-ethereum/analysis_tool_python/crawler/forked_json/2018-09-17.txt
loading  /Users/Jinyue/Documents/go-ethereum/analysis_tool_python/crawler/forked_json/2018-09-18.txt
loading  /Users/Jinyue/Documents/go-ethereum/analysis_tool_python/crawler/forked_json/2018-

loading  /Users/Jinyue/Documents/go-ethereum/analysis_tool_python/crawler/forked_json/2018-12-01.txt
loading  /Users/Jinyue/Documents/go-ethereum/analysis_tool_python/crawler/forked_json/2018-12-02.txt
loading  /Users/Jinyue/Documents/go-ethereum/analysis_tool_python/crawler/forked_json/2018-12-03.txt
loading  /Users/Jinyue/Documents/go-ethereum/analysis_tool_python/crawler/forked_json/2018-12-04.txt
loading  /Users/Jinyue/Documents/go-ethereum/analysis_tool_python/crawler/forked_json/2018-12-05.txt
loading  /Users/Jinyue/Documents/go-ethereum/analysis_tool_python/crawler/forked_json/2018-12-06.txt
loading  /Users/Jinyue/Documents/go-ethereum/analysis_tool_python/crawler/forked_json/2018-12-07.txt
loading  /Users/Jinyue/Documents/go-ethereum/analysis_tool_python/crawler/forked_json/2018-12-08.txt
loading  /Users/Jinyue/Documents/go-ethereum/analysis_tool_python/crawler/forked_json/2018-12-09.txt
loading  /Users/Jinyue/Documents/go-ethereum/analysis_tool_python/crawler/forked_json/2018-

loading  /Users/Jinyue/Documents/go-ethereum/analysis_tool_python/crawler/forked_json/2019-02-21.txt
loading  /Users/Jinyue/Documents/go-ethereum/analysis_tool_python/crawler/forked_json/2019-02-22.txt
loading  /Users/Jinyue/Documents/go-ethereum/analysis_tool_python/crawler/forked_json/2019-02-23.txt
loading  /Users/Jinyue/Documents/go-ethereum/analysis_tool_python/crawler/forked_json/2019-02-24.txt
loading  /Users/Jinyue/Documents/go-ethereum/analysis_tool_python/crawler/forked_json/2019-02-25.txt
loading  /Users/Jinyue/Documents/go-ethereum/analysis_tool_python/crawler/forked_json/2019-02-26.txt
loading  /Users/Jinyue/Documents/go-ethereum/analysis_tool_python/crawler/forked_json/2019-02-27.txt
loading  /Users/Jinyue/Documents/go-ethereum/analysis_tool_python/crawler/forked_json/2019-02-28.txt
loading  /Users/Jinyue/Documents/go-ethereum/analysis_tool_python/crawler/forked_json/2019-03-01.txt
loading  /Users/Jinyue/Documents/go-ethereum/analysis_tool_python/crawler/forked_json/2019-

loading  /Users/Jinyue/Documents/go-ethereum/analysis_tool_python/crawler/forked_json/2019-05-14.txt
loading  /Users/Jinyue/Documents/go-ethereum/analysis_tool_python/crawler/forked_json/2019-05-15.txt
loading  /Users/Jinyue/Documents/go-ethereum/analysis_tool_python/crawler/forked_json/2019-05-16.txt
loading  /Users/Jinyue/Documents/go-ethereum/analysis_tool_python/crawler/forked_json/2019-05-17.txt
loading  /Users/Jinyue/Documents/go-ethereum/analysis_tool_python/crawler/forked_json/2019-05-18.txt
loading  /Users/Jinyue/Documents/go-ethereum/analysis_tool_python/crawler/forked_json/2019-05-19.txt
loading  /Users/Jinyue/Documents/go-ethereum/analysis_tool_python/crawler/forked_json/2019-05-20.txt
loading  /Users/Jinyue/Documents/go-ethereum/analysis_tool_python/crawler/forked_json/2019-05-21.txt
loading  /Users/Jinyue/Documents/go-ethereum/analysis_tool_python/crawler/forked_json/2019-05-22.txt
loading  /Users/Jinyue/Documents/go-ethereum/analysis_tool_python/crawler/forked_json/2019-

difficulty                                          extraData  \
0  0x399985cba0ac  01040a/Geth/go1.6.2/linux (Hex:0xd78301040a844...   
1  0x39d3709c85a8               DwarfPool (Hex:0x4477617266506f6f6c)   
2  0x3984425cfd32       七彩神仙鱼 (Hex:0xe4b883e5bda9e7a59ee4bb99e9b1bc)   
3  0x39f0d1cb568c  ethermine.org (US1) (Hex:0x65746865726d696e652...   
4  0x39f815e59534  ethpool.org (ASIA1) (Hex:0x657468706f6f6c2e6f7...   

   gasLimit  gasUsed                                               hash  \
0  0x47e7c4      0x0  0xebeabb2e972e6dfe1f3cc7d6eeea5006cea363a24dd4...   
1  0x47c3d8   0x5208  0xc3a168aebfe6a19ca3063527a5eb990a56de91d74b69...   
2  0x47e7bb      0x0  0x81181666515c77a112739d09736e4db495d7b27b0143...   
3  0x47e7c4  0x19a28  0xd7aecb487cef45ab492f18e78ae41be0e3f9f8d7e617...   
4  0x47e7c4  0x4e5fe  0x16edcf526068f01be55c7200fb7dff56145a32a8dd3b...   

   mineTime                                       miner    minerName  \
0         1  0xf3b9d2c81f2b24b0fa0acaaa865b7d9ced5fc2fb  BitClubPool   
1        19  0x2a65aca4d5fc5b5c859090a6c34d164135398226  DwarfPool 1   
2         3  0x61c808d82a3ac53231750dadc13c777b59310bd9     F2Pool 1   
3        26  0xea674fdde714fd979de3edf0f56aa9716b898ec8    Ethermine   
4        49  0x4bb96091ee9d802ed039c4d1a5f6216f90f81b01    Ethpool 2   

                nonce    number  \
0  0xd20fb80e6b3e3ec1  0x204864   
1  0x8941eb20052b5b1f  0x20482a   
2  0x41e1da103c78184d  0x2047ea   
3  0xb7787292ed45fad5  0x204796   
4  0x04c95cc0042f4bda  0x20478b   

                                          parentHash reorgDepth    reward  \
0  0xd06677760086ebf7ac5d03af2a4c68b5725b4e60fad9...             5.000000   
1  0x41538a2e1ba4df21a304f995121e5c7771391c810266...             5.156691   
2  0xd0c72da348f209524053141665ae9eb29897cd3fdd9a...             5.000000   
3  0xfc7b1706682f710489f1a7467bfdbda1602f3645f180...             5.314600   
4  0x08df04e8ef851676b9faaed5768e64c93be5aabc3d6a...             5.006662   

                                          sha3Uncles   size   timestamp  \
0  0x1dcc4de8dec75d7aab85b567b6ccd41ad312451b948a...  0x21d  0x57baa20c   
1  0xebce7b8b139d85b33ff607157cbe67d6543bd991fde6...  0x49b  0x57ba9e85   
2  0x1dcc4de8dec75d7aab85b567b6ccd41ad312451b948a...  0x214  0x57ba9b6f   
3  0x4008fdcc72f27d87515b989f457405c98f21dc2a2b07...  0x87a  0x57ba9653   
4  0x1dcc4de8dec75d7aab85b567b6ccd41ad312451b948a...  0x85a  0x57ba95ad   

       totalDifficulty                         uncleReward  
0  0x2bfd60858e761620c                                   0  
1  0x2bfc8fe81090ef6cd  3.75 Ether (1 uncle at Position 0)  
2  0x2bfba976d62fcbe93                                   0  
3  0x2bfa7959e19029654  6.25 Ether (2 uncles at Position 0  
4  0x2bfa518198ca1cb80                                   0

## Load Canonical

In [11]:
canonical_df = None

for filename in sorted(os.listdir(CANONICAL_PATH)):
    if filename.endswith(".json"):
        print('loading ', CANONICAL_PATH + filename)
        new_canonical = pd.read_json(CANONICAL_PATH + filename, lines=True)
        canonical_df = pd.concat([canonical_df, new_canonical], ignore_index=True)

print(canonical_df.shape)
canonical_df.head()

loading  /Users/Jinyue/Documents/go-ethereum/records/blocks/canonical/json/6830000.json
loading  /Users/Jinyue/Documents/go-ethereum/records/blocks/canonical/json/6840000.json
loading  /Users/Jinyue/Documents/go-ethereum/records/blocks/canonical/json/6850000.json
loading  /Users/Jinyue/Documents/go-ethereum/records/blocks/canonical/json/6860000.json
loading  /Users/Jinyue/Documents/go-ethereum/records/blocks/canonical/json/6870000.json
loading  /Users/Jinyue/Documents/go-ethereum/records/blocks/canonical/json/6880000.json
loading  /Users/Jinyue/Documents/go-ethereum/records/blocks/canonical/json/6890000.json
loading  /Users/Jinyue/Documents/go-ethereum/records/blocks/canonical/json/6900000.json
loading  /Users/Jinyue/Documents/go-ethereum/records/blocks/canonical/json/6910000.json
loading  /Users/Jinyue/Documents/go-ethereum/records/blocks/canonical/json/6920000.json
loading  /Users/Jinyue/Documents/go-ethereum/records/blocks/canonical/json/6930000.json
loading  /Users/Jinyue/Documents

loading  /Users/Jinyue/Documents/go-ethereum/records/blocks/canonical/json/7770000.json
loading  /Users/Jinyue/Documents/go-ethereum/records/blocks/canonical/json/7780000.json
loading  /Users/Jinyue/Documents/go-ethereum/records/blocks/canonical/json/7790000.json
loading  /Users/Jinyue/Documents/go-ethereum/records/blocks/canonical/json/7800000.json
loading  /Users/Jinyue/Documents/go-ethereum/records/blocks/canonical/json/7810000.json
loading  /Users/Jinyue/Documents/go-ethereum/records/blocks/canonical/json/7820000.json
loading  /Users/Jinyue/Documents/go-ethereum/records/blocks/canonical/json/7830000.json
loading  /Users/Jinyue/Documents/go-ethereum/records/blocks/canonical/json/7840000.json
loading  /Users/Jinyue/Documents/go-ethereum/records/blocks/canonical/json/7850000.json
loading  /Users/Jinyue/Documents/go-ethereum/records/blocks/canonical/json/7860000.json
loading  /Users/Jinyue/Documents/go-ethereum/records/blocks/canonical/json/7870000.json
loading  /Users/Jinyue/Documents

difficulty                                   extraData  gasLimit  \
0  0x88622c2bfedab  0x737061726b706f6f6c2d6574682d636e2d687a32   8000029   
1  0x88723a734d445                      0x70616e64615f706f6f6c   7994877   
2  0x88713a734d445                  0x6e616e6f706f6f6c2e6f7267   8000029   
3  0x88814a9ca0d86                        0x4477617266506f6f6c   8000000   
4  0x886f3cb50a372            0xe4b883e5bda9e7a59ee4bb99e9b1bc   8000000   

   gasUsed                                               hash  \
0  7999694  0xa74abd1fe708da69f311447bde33193e9b23af3a6eec...   
1   742283  0xd6a9a861552586be4f4ecbaf14754d142bb1887e67f8...   
2  1815372  0xc8958f7838145975ab421e3a0c08a4eeab955233fd63...   
3  7985081  0x423d96a9aaf78b780bd0bb93a25a45a57260f16b1106...   
4  7610034  0xbee381dabfe10ec32d6d28cd34c1c7200d0593f84d0d...   

                                           logsBloom mineTime  \
0  0x9c025060a0e001869202029430848d1220a5041ca202...            
1  0x08200000000000000000040088000000000000410400...            
2  0x00080020840000000800828084008008400105011400...            
3  0x048c24e004000a88a44880560800802100a0b0802810...            
4  0x087200020004058e2420101740a79125002048118111...            

                                        miner minerName  \
0  0x5a0b54d5dc17e0aadc383d2db43b0a0d3e029c4c             
1  0x2a5994b501e6a560e727b6c2de5d856396aadd38             
2  0x52bc44d5378309ee2abf1539bf71de1b7d7be3b5             
3  0x2a65aca4d5fc5b5c859090a6c34d164135398226             
4  0x829bd824b016326a401d083b33d092293333a830             

                                             mixHash  ...   size  \
0  0xb63620abe8b24f69f91e12e2b81e1af21d77085a0eaf...  ...  30787   
1  0x1d234210b6a1dc7a99337cf124e72648334463068eb4...  ...   3598   
2  0x890f089c4711a8a1ca870371d60d92b0ce679bead3e6...  ...   7253   
3  0xbf1762fa9566719122e61dc0e55656db13e787b8dcb9...  ...  19764   
4  0x0d8a7763a52dadfc610a969aa458e988a85f2541b60b...  ...  28231   

                                           stateRoot           timestamp  \
0  0x02bb13e7b92e4d84e6befca63e4773a444c184d12f7e... 2018-12-07 02:03:14   
1  0xc7e8aafe7dcf3d5c592c81dc04432b653d8cce14fb96... 2018-12-07 02:02:51   
2  0x3dfe8d6556f7ff318bb2e8a9d2c30432fbf466714fef... 2018-12-07 02:02:40   
3  0xa96329d85108e557661b3415ec89e87b64d7cc46523b... 2018-12-07 02:02:22   
4  0x713556c78986dc2b99b0006bb5f244785030e8ffaed1... 2018-12-07 02:02:17   

         totalDifficulty                                       transactions  \
0  0x1bafafae6b4f5918e7f  [0x3dd0debcd89a3752326d31f55aedd7e69fa143645ff...   
1  0x1bafaf2609232d1a0d4  [0x9f564c885eb3e94655f000732df45e42a02854435a7...   
2  0x1bafae9d96e8b9ccc8f  [0x73c2462415d5f86f2763f7d976e898bc0e006feb7ab...   
3  0x1bafae1525ae467f84a  [0xbf457bd1d4fada3194f7da6804a76f62771d140f7bb...   
4  0x1bafad8ca463a9deac4  [0x8da1ec8b939508339097656c2568f8449d971d72408...   

                                    transactionsRoot txNum  uncleNum  \
0  0x35961cc3036b682f2cdc2773d967cd8fdb93d4aec3a9...   183         0   
1  0xb89bad1ef046cf73d47a9c68d4978cbc4c9f5e0a6cd6...    15         0   
2  0xb409b8445c8f6236914e124c73046536268a874b793f...    46         0   
3  0x2502639b0d30d0af28ea1be3b37aed8550fcadbb5c1f...   127         0   
4  0xb3e23d6ae0a85c1c86a67a9396a8b2ab8c69c5c21fd4...   194         0   

  uncleReward uncles  
0                 []  
1                 []  
2                 []  
3                 []  
4                 []  

[5 rows x 27 columns]

## 1. LogsBloom: Check how much percentage and how many instances have this feature 

### Broadcast

In [16]:
broadcast_df.logsBloom.isnull().values.any()

False

### Uncles

In [17]:
uncle_df.logsBloom.isnull().values.any()

False

### Forked

In [19]:
forked_df.logsBloom.isnull().values.any()

AttributeError: 'DataFrame' object has no attribute 'logsBloom'

In [20]:
forked_df.columns

Index(['difficulty', 'extraData', 'gasLimit', 'gasUsed', 'hash', 'mineTime',
       'miner', 'minerName', 'nonce', 'number', 'parentHash', 'reorgDepth',
       'reward', 'sha3Uncles', 'size', 'timestamp', 'totalDifficulty',
       'uncleReward'],
      dtype='object')

#### Forked blocks dataset does not have the feature logsBloom

### Canonical 

In [21]:
canonical_df.columns

Index(['difficulty', 'extraData', 'gasLimit', 'gasUsed', 'hash', 'logsBloom',
       'mineTime', 'miner', 'minerName', 'mixHash', 'nonce', 'number',
       'parentHash', 'receiptsRoot', 'reorgDepth', 'reward', 'sha3Uncles',
       'size', 'stateRoot', 'timestamp', 'totalDifficulty', 'transactions',
       'transactionsRoot', 'txNum', 'uncleNum', 'uncleReward', 'uncles'],
      dtype='object')

In [22]:
canonical_df.logsBloom.isnull().values.any()

False

### So, only the forked dataset does not have the feature logsBloom. Let's check if each miner only uses one logsBloom.

In [25]:
uncle_df.groupby(['miner', 'logsBloom']).size()

miner                                       logsBloom                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                         
0x00192fb10df37c9fb26829eb2cc623cd1bf599e8  0x00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000

In [27]:
canonical_df.groupby(['miner', 'logsBloom']).size()

miner                                       logsBloom                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                         
0x000b800ab6000bc8e5c02423723bbb168720530a  0x000001801120890c816280400040200000008000040044000168a00002010108088000400820300004c80102101008000004100002020100000010084021400800120802870300800989024860040840480c44418c0000900000202b0480004011018407c2006085902008010800188048089000041100041204981040015010008100800002000002000288240401082040c0300100120080061102004040000260200008082404000110d0300030220008002000918001082040901800414820120202404

In [28]:
broadcast_df.groupby(['miner', 'logsBloom']).size()

miner                                       logsBloom                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                       
                                            000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000

### We can find that for each miner, it uses many different logsBlooms. 

### Compare canonical, uncle, and broadcast datasets with forked dataset, if the hash value matches, append the logsBloom to the forked dataset.

In [38]:
forked_df.shape

(94483, 18)

In [46]:
bc_uncle_can_df = pd.concat([broadcast_df, uncle_df])
bc_uncle_can_df = pd.concat([bc_uncle_can_df, canonical_df])
bc_uncle_can_df.head()

/Users/Jinyue/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:1: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version
of pandas will change to not sort by default.

To accept the future behavior, pass 'sort=False'.

To retain the current behavior and silence the warning, pass 'sort=True'.

  """Entry point for launching an IPython kernel.
/Users/Jinyue/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:2: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version
of pandas will change to not sort by default.

To accept the future behavior, pass 'sort=False'.

To retain the current behavior and silence the warning, pass 'sort=True'.

  


difficulty extra extraData  gasLimit   gasUsed  \
0  2119842492649275   NaN            0x7a20b0  0x484e2d   
1  2119842492649275   NaN            0x7a20b0  0x484e2d   
2  2119843029520187   NaN            0x7a1d46  0x39bfb3   
3  2119843029520187   NaN            0x7a1d46  0x39bfb3   
4                     NaN            0x7a1d46  0x39bfb3   

                                                hash  \
0  0x869d968cbd5404f81d8fa750eb4672f4752db1979474...   
1  0x869d968cbd5404f81d8fa750eb4672f4752db1979474...   
2  0x86c7e5fd62d71430be8292ac909df6a1b3730aae7e07...   
3  0x86c7e5fd62d71430be8292ac909df6a1b3730aae7e07...   
4  0x86c7e5fd62d71430be8292ac909df6a1b3730aae7e07...   

                                           logsBloom mineTime  \
0  0221218a04322008408904084268000981003008400000...            
1  0221218a04322008408904084268000981003008400000...            
2  000050000000002a845c00002408901421c230c2011020...            
3  000050000000002a845c00002408901421c230c2011020...            
4  000050000000002a845c00002408901421c230c2011020...            

                                        miner minerName  ... stateRoot  \
0  0x5A0b54D5dc17e0AadC383d2db43B0a0D3E029c4c            ...             
1  0x5A0b54D5dc17e0AadC383d2db43B0a0D3E029c4c            ...             
2  0x5A0b54D5dc17e0AadC383d2db43B0a0D3E029c4c            ...             
3  0x5A0b54D5dc17e0AadC383d2db43B0a0D3E029c4c            ...             
4  0x5A0b54D5dc17e0AadC383d2db43B0a0D3E029c4c            ...             

            timestamp totalDifficulty  transactions transactionsRoot txNum  \
0 2019-07-08 22:59:50                                                   65   
1 2019-07-08 22:59:50                                                   65   
2 2019-07-08 22:59:59                                                   58   
3 2019-07-08 22:59:59                                                   58   
4 2019-07-08 22:59:59                                                   58   

  uncleHash uncleNum uncleReward  \
0       NaN      0.0               
1       NaN      0.0               
2       NaN      0.0               
3       NaN      0.0               
4       NaN      0.0               

                                              uncles  
0  [0x1dcc4de8dec75d7aab85b567b6ccd41ad312451b948...  
1  [0x1dcc4de8dec75d7aab85b567b6ccd41ad312451b948...  
2  [0x1dcc4de8dec75d7aab85b567b6ccd41ad312451b948...  
3  [0x1dcc4de8dec75d7aab85b567b6ccd41ad312451b948...  
4  [0x1dcc4de8dec75d7aab85b567b6ccd41ad312451b948...  

[5 rows x 32 columns]

In [47]:
full_mergedStuff = pd.merge(forked_df, bc_uncle_can_df, on=['hash'], how='inner')
full_mergedStuff.shape

(40559, 49)

In [49]:
print(40559/94483*100, '% forked blocks can have logsBloom.')

42.927299090841736 % forked blocks can have logsBloom.


### 40559 blocks and 42.93% of the forked blocks can have logsBloom

## 2. MineTime: Check how much percentage and how many instances have this feature